crates/storage/provider/src/providers/snapshot/manager.rs

```diff
@@ -223,39 +219,90 @@ impl SnapshotProvider {
             }
             let tx_start = snapshots_rev_iter.peek().map(|(tx_end, _)| *tx_end + 1).unwrap_or(0);
             if tx_start <= tx {
-                return Some((block_range.clone(), tx_start..=*tx_end))
+                return Some(find_fixed_range(BLOCKS_PER_SNAPSHOT, *block_range.end()))
             }
         }
         None
     }
 
-    /// Updates the inner transaction and block index
-    pub fn update_index(&self) -> ProviderResult<()> {
-        let mut block_index = self.snapshots_block_index.write();
+    /// Updates the inner transaction and block index alongside the internal `self.map`.
+    ///
+    /// Any entry higher than `segment_max_block` will be deleted from the previous structures.
+    pub fn update_index(
+        &self,
+        segment: SnapshotSegment,
+        segment_max_block: Option<BlockNumber>,
+    ) -> ProviderResult<()> {
+        let mut max_block = self.snapshots_max_block.write();
+        let mut tx_index = self.snapshots_tx_index.write();
+
+        match segment_max_block {
+            Some(segment_max_block) => {
+                // Update the max block for the segment
+                max_block.insert(segment, segment_max_block);
+                let fixed_range = find_fixed_range(BLOCKS_PER_SNAPSHOT, segment_max_block);
+
+                let jar = NippyJar::<SegmentHeader>::load(
+                    &self.path.join(segment.filename(&fixed_range)),
+                )?;
+
+                // Updates the tx index by first removing all entries which have a higher
+                // block_start than our current static file.
+                if let Some(tx_range) = jar.user_header().tx_range() {
+                    let tx_end = *tx_range.end();
+                    let block_range = jar.user_header().block_range().clone();
+
+                    tx_index
+                        .entry(segment)
+                        .and_modify(|index| {
+                            index.retain(|_, range| range.start() < fixed_range.start());
+                            index.insert(tx_end, block_range.clone());
+                        })
+                        .or_insert_with(|| BTreeMap::from([(tx_end, block_range)]));
+                }
+
+                // Update the cached provider.
+                self.map.insert((*fixed_range.end(), segment), LoadedJar::new(jar)?);
+
+                // Delete any cached provider that no longer has an associated jar.
```

Comment:
it has an associated jar, but it was updated and outdated now, right?

Reply:
or is it for the case when we prune blocks from the jar?

Reply:
> or is it for the case when we prune blocks from the jar?

yes,  when the pruning goes across jars, in such a way that the `update_index` is called in a lower block than the initial (and now deleted) jar